src/main/scala/is/hail/sparkextras/ContextSampledRDD.scala

```diff
@@ -0,0 +1,47 @@
+package is.hail.sparkextras
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.util.random.RandomSampler
+
+import java.util.Random
+
+import scala.reflect.ClassTag
+
+class SeededPartition(
+  val prev: Partition,
+  val seed: Long
+) extends Partition with Serializable {
+  override val index: Int = prev.index
+}
+
+// FIXME: this doesn't feel like it should be necessary, can't we do this by
+// zipPartitions with a seed and then mapPartitions?
+class ContextSampledRDD[C, T: ClassTag, U: ClassTag](
+  prev: RDD[C => Iterator[T]],
+  sampler: RandomSampler[T, U],
+  @transient private val seed: Long
+) extends RDD[C => Iterator[U]](prev) {
+
+  @transient override val partitioner = prev.partitioner
+
+  override def getPartitions: Array[Partition] = {
+    val random = new Random(seed)
+    prev.partitions.map(x => new SeededPartition(x, random.nextLong()))
+  }
+
+  override def getPreferredLocations(split: Partition): Seq[String] =
```

Comment:
do you need this? Passing RDD(prev) should create a narrow dependency on prev's partitions, and I think narrow dependencies hook up with Spark's preferred location stuff

Reply:
It looks like it is necessary.  From RDD:

 >   protected def getPreferredLocations(split: Partition): Seq[String] = Nil

The default locations logic doesn't look at the dependencies.

Reply:
```
private[spark] class MapPartitionsRDD[U: ClassTag, T: ClassTag](
    var prev: RDD[T],
    f: (TaskContext, Int, Iterator[T]) => Iterator[U],  // (TaskContext, partition index, iterator)
    preservesPartitioning: Boolean = false)
  extends RDD[U](prev) {

  override val partitioner = if (preservesPartitioning) firstParent[T].partitioner else None

  override def getPartitions: Array[Partition] = firstParent[T].partitions

  override def compute(split: Partition, context: TaskContext): Iterator[U] =
    f(context, split.index, firstParent[T].iterator(split, context))

  override def clearDependencies() {
    super.clearDependencies()
    prev = null
  }
}
```

Reply:
given this code, I think we must be misunderstanding how it works. no?

Reply:
meaning that if you're supposed to set preferred locations everywhere, spark would definitely do it here.

Reply:
I'm at a complete loss.

Reply:
Ah, it is not necessary.  DagScheduler.getPreferredLocsInternal handles the case of narrow dependencies.